devdoc/channel.md

```diff
@@ -0,0 +1,267 @@
+
+**SRS_CHANNEL_43_029: [** If the current `state` is `CHANNEL_PULLED`, `channel_push` shall: **]**
+
+- **SRS_WAITER_43_001: [** store `pull_callback` and `pull_context` given to `channel_pull`. **]**
+
+- **SRS_CHANNEL_43_043: [** call `interlocked_compare_exchange` to increment the `EPOCH` and change the state to `CHANNEL_READY`. **]**
+
+- **SRS_CHANNEL_43_030: [** call the stored `pull_callback` with the stored `pull_context`, `data` as the given `data` and `result` as `CHANNEL_RESULT_OK`. **]**
+
+- **SRS_CHANNEL_43_031: [** call the given `push_callback` with `push_context` as the given `push_context` and `result` as `CHANNEL_RESULT_OK`. **]**
+
+- **SRS_CHANNEL_43_055: [** set `out_op_push` to `NULL` by calling `THANDLE_ASSIGN`. **]**
+
+- **SRS_CHANNEL_43_032: [** return `CHANNEL_RESULT_SYNC`. **]**
+
+**SRS_CHANNEL_43_033: [** If the current state is `CHANNEL_READY`, `channel_push` shall: **]**
+
+- **SRS_CHANNEL_43_065: [** set the state to `CHANNEL_PUSHING` by calling `interlocked_compare_exchange`. **]**
+
+- **SRS_CHANNEL_43_034: [** call `async_op_create` with `cancel_push` as the cancel function. **]**
+
+- **SRS_CHANNEL_43_035: [** store the given `push_callback` and the given `push_context` in the `channel`. **]**
+
+- **SRS_CHANNEL_43_036: [** store the given `data` in the `channel` by calling `THANDLE_INITIALIZE`. **]**
+
+- **SRS_CHANNEL_43_037: [** store `state` in the created `THANDLE(ASYNC_OP)`. **]**
+
+- **SRS_CHANNEL_43_038: [** store the given `channel` in the created `THANDLE(ASYNC_OP)` by calling `THANDLE_INITIALIZE`. **]**
+
+- **SRS_CHANNEL_43_039: [** store the created `THANDLE(ASYNC_OP)` in `out_op_push`by calling `THANDLE_INITIALIZE_MOVE`. **]**
+
+set the `state` of the given `channel` to `CHANNEL_PUSHED` by calling `interlocked_exchange`.
+
+signal waiting threads by calling `wake_by_address_all`.
+
+- **SRS_CHANNEL_43_040: [** return `CHANNEL_RESULT_ASYNC`. **]**
+
+**SRS_CHANNEL_43_041: [** If there are any failures, `channel_push` shall fail and return `CHANNEL_RESULT_ERROR`. **]**
+
+
+### cancel_push
+```c
+static void cancel_push(void* context)
+```
+
+`cancel_push` is the cancel function given to `async_op_create` when `channel_push` is called.
+
+**SRS_CHANNEL_43_049: [** If the current `EPOCH` and `state` of the `channel` is the same as the initial `EPOCH` and `state` of the `channel` when `channel_push` was called, `cancel_push` shall: **]**
+
+- **SRS_CHANNEL_43_050: [** increment the `EPOCH` and set the `state` of the `channel` stored in the given `context` to `CHANNEL_READY` by calling `interlocked_compare_exchange`. **]**
+
+- **SRS_CHANNEL_43_051: [** call the `push_callback` given to `channel_push` with `push_context` as the `push_context` given to `channel_push` and `result` as `CHANNEL_RESULT_CANCELLED`. **]**
+
+- **SRS_CHANNEL_43_052: [** release the reference to `data` stored in `channel` by calling `THANDLE_ASSIGN`. **]**
```

Comment:
>SRS_CHANNEL_43_052 [](http://example.com/codeflow?start=4&length=18)

I will be waiting because the interlocked comments and the state management in general.